File: apps/student/src/main/java/com/instructure/student/util/ModuleUtility.java
/*
 * Copyright (C) 2016 - present Instructure, Inc.
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, version 3 of the License.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

package com.instructure.student.util;

import android.net.Uri;
import android.os.Bundle;

import com.instructure.canvasapi2.models.Course;
import com.instructure.canvasapi2.models.ModuleItem;
import com.instructure.canvasapi2.models.ModuleObject;
import com.instructure.canvasapi2.utils.APIHelper;
import com.instructure.interactions.router.Route;
import com.instructure.student.fragment.DiscussionDetailsFragment;
import com.instructure.student.fragment.FileDetailsFragment;
import com.instructure.student.fragment.InternalWebviewFragment;
import com.instructure.student.fragment.MasteryPathLockedFragment;
import com.instructure.student.fragment.MasteryPathSelectionFragment;
import com.instructure.student.fragment.ModuleQuizDecider;
import com.instructure.student.fragment.PageDetailsFragment;
import com.instructure.student.mobius.assignmentDetails.ui.AssignmentDetailsFragment;

import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;
import java.util.Date;

import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

public class ModuleUtility {

    ///////////////////////////////////////////////////////////////////////////
    // Fragment Creation
    ///////////////////////////////////////////////////////////////////////////

    public static Fragment getFragment(ModuleItem moduleItem, final Course course, ModuleObject moduleObject) {
        if(moduleItem.getType() == null) {
            return null;
        }

        //deal with files
        if(moduleItem.getType().equals("File")) {
            String url = moduleItem.getUrl();
            //we just want the end of the api, not the domain

            Bundle bundle;
            url = removeDomain(url);

            if(moduleObject == null){
                return FileDetailsFragment.newInstance(FileDetailsFragment.makeRoute(course, url));
            } else {
                return FileDetailsFragment.newInstance(FileDetailsFragment.makeRoute(course, moduleObject, moduleItem.getId(), url));
            }
        }

        //deal with pages
        if (moduleItem.getType().equals("Page")) {
            return PageDetailsFragment.newInstance(PageDetailsFragment.makeRoute(course, moduleItem.getTitle(), moduleItem.getPageUrl()));
        }

        //deal with assignments
        if(moduleItem.getType().equals("Assignment")) {
            return AssignmentDetailsFragment.newInstance(AssignmentDetailsFragment.makeRoute(course, getAssignmentId(moduleItem, course)));
        }

        //deal with external urls
        if((moduleItem.getType().equals("ExternalUrl") || moduleItem.getType().equals("ExternalTool"))) {

            Uri uri =  Uri.parse(moduleItem.getHtmlUrl()).buildUpon().appendQueryParameter("display", "borderless").build();

            Route route = InternalWebviewFragment.Companion.makeRoute(course, uri.toString(), moduleItem.getTitle(), true, true, true);
            return InternalWebviewFragment.Companion.newInstance(route);
        }

        //don't do anything with headers, they're just dividers so we don't show them here.
        if((moduleItem.getType().equals("SubHeader"))) {
            return null;
        }

        // Quizzes
        if (moduleItem.getType().equals("Quiz")) {
            String apiURL = moduleItem.getUrl();
            apiURL = removeDomain(apiURL);
            return ModuleQuizDecider.newInstance(ModuleQuizDecider.makeRoute(course, moduleItem.getHtmlUrl(), apiURL));
        }

        //Discussions
        if(moduleItem.getType().equals("Discussion")) {
            Route route = getDiscussionRoute(moduleItem, course);
            return DiscussionDetailsFragment.newInstance(route);
        }

        if(moduleItem.getType().equals("Locked")) {
            Route route = MasteryPathLockedFragment.makeRoute(moduleItem.getTitle());
            return MasteryPathLockedFragment.newInstance(route);
        }

        if(moduleItem.getType().equals("ChooseAssignmentGroup")) {
            Route route = MasteryPathSelectionFragment.makeRoute(course, moduleItem.getMasteryPaths(), moduleObject.getId(), moduleItem.getMasteryPathsItemId());
            return MasteryPathSelectionFragment.newInstance(route);
        }
        //return null if there is a type we don't handle yet
        return null;
    }

    ///////////////////////////////////////////////////////////////////////////
    // Bundle Creation
    ///////////////////////////////////////////////////////////////////////////

    @Nullable
    private static String getPageName(ModuleItem moduleItem) {
        //get the pageName from the url
        String url = moduleItem.getUrl();
        String pageName = null;
        if(url != null) {
            if(url.contains("wiki/")) {
                int index = url.indexOf("wiki/");
                if(index != -1) {
                    index += 5;
                    pageName = getPageName(url, index);
                }
            }
            else if(url.contains("pages/")) {
                int index = url.indexOf("pages/");
                if(index != -1) {
                    index += 6;
                    pageName = getPageName(url, index);
                }
            }
        }

        return pageName;
    }

    private static long getAssignmentId(ModuleItem moduleItem, Course course) {
        //get the assignment id from the url
        String url = moduleItem.getUrl();
        long assignmentId = 0;
        if(url.contains("assignments")) {
            int index = url.indexOf("assignments/");
            if(index != -1) {
                index += 12;
                assignmentId = getIdFromUrl(url, index);
            }
        }
        return assignmentId;
    }

    private static Route getDiscussionRoute(ModuleItem moduleItem, Course course) {
        // Get the topic id from the url
        String url = moduleItem.getUrl();
        long topicId = 0;
        if(url.contains("discussion_topics")) {
            int index = url.indexOf("discussion_topics/");
            if(index != -1) {
                index += 18;
                topicId = getIdFromUrl(url, index);
            }
        }
        return DiscussionDetailsFragment.makeRoute(course, topicId, null);
    }

    ///////////////////////////////////////////////////////////////////////////
    // Helpers
    ///////////////////////////////////////////////////////////////////////////

    public static boolean isGroupLocked(ModuleObject module) {
        // Currently the state for the group says "Locked" until the user visits the modules online, so we need
        // a different way to determine if it's Locked
        boolean isLocked = false;
        Date curDate = new Date();
        boolean isDatePassed;
        // Check if we're passed the unlock date
        if (module.getUnlockDate() != null && curDate.before(module.getUnlockDate())) return true;
        else isDatePassed = true;

        long[] ids = module.getPrerequisiteIds();
        // Is the unlock date is in the past or the state == Locked AND there are prerequisites
        if (ids != null && (isDatePassed || (ids.length > 0 && ids[0] != 0)) && module.getState() != null &&
                module.getState().equals(ModuleObject.State.Locked.getApiString())) {
            isLocked = true;
        }

        return isLocked;
    }

    private static String removeDomain(String url) {
        //strip off the domain and protocol
        int index = 0;
        String prefix = "/api/v1/";
        index = url.indexOf(prefix);
        if(index != -1) {
            url = url.substring(index + prefix.length());
        }
        return url;
    }

    private static long getIdFromUrl(String url, int index) {
        long assignmentId;
        int endIndex = url.indexOf("/", index);
        if(endIndex != -1) {
            assignmentId = Long.parseLong(APIHelper.INSTANCE.expandTildeId(url.substring(index, endIndex)));
        } else {
            assignmentId = Long.parseLong(APIHelper.INSTANCE.expandTildeId(url.substring(index)));
        }
        return assignmentId;
    }

    private static String getPageName(String url, int index) {
        String pageName;
        int endIndex = url.indexOf("/", index);
        if(endIndex != -1) {
            pageName = url.substring(index, endIndex);
        } else {
            pageName = url.substring(index);
        }

        //decode the page name in case there are special characters in the name (like {}<>|`)
        try {
            pageName = URLDecoder.decode(pageName, "UTF-8");
        }
        catch (UnsupportedEncodingException e) {
            LoggingUtility.LogConsole(e.getMessage());
        }

        return pageName;
    }
}

File: apps/student/src/main/java/com/instructure/student/features/modules/util/ModuleUtility.kt
/*
 * Copyright (C) 2016 - present Instructure, Inc.
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, version 3 of the License.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
package com.instructure.student.features.modules.util

import android.content.Context
import android.net.Uri
import androidx.fragment.app.Fragment
import com.instructure.canvasapi2.models.Course
import com.instructure.canvasapi2.models.ModuleItem
import com.instructure.canvasapi2.models.ModuleObject
import com.instructure.canvasapi2.models.Tab
import com.instructure.canvasapi2.utils.APIHelper.expandTildeId
import com.instructure.canvasapi2.utils.findWithPrevious
import com.instructure.canvasapi2.utils.isLocked
import com.instructure.interactions.router.Route
import com.instructure.pandautils.features.discussion.details.DiscussionDetailsWebViewFragment
import com.instructure.student.R
import com.instructure.pandautils.features.assignments.details.AssignmentDetailsFragment
import com.instructure.pandautils.features.assignments.details.AssignmentDetailsFragment.Companion.makeRoute
import com.instructure.student.features.discussion.details.DiscussionDetailsFragment
import com.instructure.student.features.discussion.details.DiscussionDetailsFragment.Companion.makeRoute
import com.instructure.student.features.files.details.FileDetailsFragment
import com.instructure.student.features.modules.progression.LockedModuleItemFragment
import com.instructure.student.features.modules.progression.ModuleQuizDecider
import com.instructure.student.features.modules.progression.NotAvailableOfflineFragment
import com.instructure.student.features.pages.details.PageDetailsFragment
import com.instructure.student.fragment.InternalWebviewFragment
import com.instructure.student.fragment.InternalWebviewFragment.Companion.makeRoute
import com.instructure.student.fragment.MasteryPathSelectionFragment
import com.instructure.student.fragment.MasteryPathSelectionFragment.Companion.makeRoute
import java.util.Date

object ModuleUtility {
    fun getFragment(
        item: ModuleItem,
        course: Course,
        moduleObject: ModuleObject?,
        navigatedFromModules: Boolean,
        isOnline: Boolean,
        syncedTabs: Set<String>,
        syncedFileIds: List<Long>,
        context: Context
    ): Fragment? = when (item.type) {
        "Page" -> PageDetailsFragment.newInstance(PageDetailsFragment.makeRoute(course, item.title, item.pageUrl, navigatedFromModules))
        "Assignment" -> {
            createFragmentWithOfflineCheck(isOnline, course, item, syncedTabs, context, setOf(Tab.ASSIGNMENTS_ID, Tab.GRADES_ID, Tab.SYLLABUS_ID)) {
                AssignmentDetailsFragment.newInstance(makeRoute(course, getAssignmentId(item)))
            }
        }
        "Discussion" -> {
            createFragmentWithOfflineCheck(isOnline, course, item, syncedTabs, context, setOf(Tab.DISCUSSIONS_ID)) {
                if (isOnline) {
                    DiscussionDetailsWebViewFragment.newInstance(getDiscussionRedesignRoute(item, course))
                } else {
                    DiscussionDetailsFragment.newInstance(getDiscussionRoute(item, course))
                }
            }
        }
        "Locked" -> LockedModuleItemFragment.newInstance(LockedModuleItemFragment.makeRoute(course, item.title!!, item.moduleDetails?.lockExplanation ?: ""))
        "SubHeader" -> null // Don't do anything with headers, they're just dividers so we don't show them here.
        "Quiz" -> {
            val apiURL = removeDomain(item.url)
            ModuleQuizDecider.newInstance(ModuleQuizDecider.makeRoute(course, item.htmlUrl!!, apiURL!!, item.contentId))
        }
        "ChooseAssignmentGroup" -> {
            createFragmentWithOfflineCheck(isOnline, course, item, syncedTabs, context) {
                val route = makeRoute(course, item.masteryPaths!!, moduleObject!!.id, item.masteryPathsItemId)
                MasteryPathSelectionFragment.newInstance(route)
            }
        }
        "ExternalUrl", "ExternalTool" -> {
            if (item.isLocked()) {
                LockedModuleItemFragment.newInstance(LockedModuleItemFragment.makeRoute(course, item.title!!, item.moduleDetails?.lockExplanation ?: ""))
            } else {
                createFragmentWithOfflineCheck(isOnline, course, item, syncedTabs, context) {
                    val uri = Uri.parse(item.htmlUrl).buildUpon().appendQueryParameter("display", "borderless").build()
                    val route = makeRoute(course, uri.toString(), item.title!!, true, true, true)
                    InternalWebviewFragment.newInstance(route)
                }
            }
        }
        "File" -> { // TODO Handle offline availability after files sync
            createFileDetailsFragmentWithOfflineCheck(isOnline, course, item, moduleObject, syncedFileIds, context)
        }
        else -> null
    }

    private fun createFragmentWithOfflineCheck(
        isOnline: Boolean,
        course: Course,
        item: ModuleItem,
        syncedTabs: Set<String>,
        context: Context,
        tabs: Set<String> = emptySet(),
        creationBlock: () -> Fragment?
    ): Fragment? {
        return if (isOnline || tabs.any { syncedTabs.contains(it) }) {
            creationBlock()
        } else {
            val descriptionResource = if (tabs.isEmpty()) R.string.notAvailableOfflineDescription else R.string.notAvailableOfflineDescriptionForTabs
            NotAvailableOfflineFragment.newInstance(NotAvailableOfflineFragment.makeRoute(course, item.title, context.getString(descriptionResource)))
        }
    }

    private fun createFileDetailsFragmentWithOfflineCheck(
        isOnline: Boolean,
        course: Course,
        item: ModuleItem,
        moduleObject: ModuleObject?,
        syncedFiles: List<Long>,
        context: Context,
    ): Fragment? {
        return if (isOnline || syncedFiles.contains(item.contentId)) {
            val url = removeDomain(item.url)
            if (moduleObject == null) {
                FileDetailsFragment.newInstance(
                    FileDetailsFragment.makeRoute(
                        course,
                        url!!,
                        item.contentId
                    )
                )
            } else {
                FileDetailsFragment.newInstance(
                    FileDetailsFragment.makeRoute(
                        course,
                        moduleObject,
                        item.id,
                        url!!,
                        item.contentId
                    )
                )
            }
        } else {
            NotAvailableOfflineFragment.newInstance(NotAvailableOfflineFragment.makeRoute(course, item.title, context.getString(R.string.notAvailableOfflineDescriptionForTabs)))
        }
    }

    fun isGroupLocked(module: ModuleObject?): Boolean {
        // NOTE: The state of the group is is "Locked" until the user visits the modules online
        // Check if the unlock date has passed
        if (module?.unlockDate?.after(Date()) == true) return true

        // Check if the state is Locked AND there are prerequisites
        return module?.prerequisiteIds != null && module.state == ModuleObject.State.Locked.apiString
    }

    private fun getAssignmentId(moduleItem: ModuleItem): Long {
        // Get the assignment id from the url
        return Uri.parse(moduleItem.url).pathSegments
            .findWithPrevious { previous, _ -> previous == "assignments" }
            ?.let { expandTildeId(it) }
            ?.toLongOrNull() ?: 0
    }

    private fun getDiscussionRoute(moduleItem: ModuleItem, course: Course): Route {
        // Get the topic id from the url
        val topicId = Uri.parse(moduleItem.url).pathSegments
            .findWithPrevious { previous, _ -> previous == "discussion_topics" }
            ?.let { expandTildeId(it) }
            ?.toLongOrNull() ?: 0
        return makeRoute(course, topicId, null)
    }

    private fun getDiscussionRedesignRoute(moduleItem: ModuleItem, course: Course): Route {
        // Get the topic id from the url
        val topicId = Uri.parse(moduleItem.url).pathSegments
            .findWithPrevious { previous, _ -> previous == "discussion_topics" }
            ?.let { expandTildeId(it) }
            ?.toLongOrNull() ?: 0
        return DiscussionDetailsWebViewFragment.makeRoute(course, topicId)
    }

    /** Strips off the domain and protocol */
    private fun removeDomain(url: String?): String? = url?.substringAfter("/api/v1/")
}
